File: ESVBible/Services/SearchService.swift
import Foundation

@Observable
class SearchService {
    /// A search result with location and text for display.
    struct VerseResult: Identifiable, Equatable {
        let book: String
        let chapter: Int
        let verse: Int
        let text: String  // looked up from BibleStore at query time
        var id: String { "\(book).\(chapter).\(verse)" }
    }

    /// Internal struct matching the JSON index format (no text).
    private struct IndexEntry: Codable {
        let book: String
        let chapter: Int
        let verse: Int
    }

    private var index: [String: [IndexEntry]] = [:]
    private var isLoaded = false

    init() {}

    /// Test-only initializer with a pre-built index.
    init(index: [String: [(book: String, chapter: Int, verse: Int)]]) {
        self.index = index.mapValues { entries in
            entries.map { IndexEntry(book: $0.book, chapter: $0.chapter, verse: $0.verse) }
        }
        self.isLoaded = true
    }

    /// Load the index from the bundled JSON file.
    func loadIfNeeded() {
        guard !isLoaded else { return }
        isLoaded = true

        guard let url = Bundle.main.url(forResource: "search_index", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let decoded = try? JSONDecoder().decode([String: [IndexEntry]].self, from: data) else {
            print("Error: Could not load or parse search_index.json from the app bundle.")
            return
        }
        self.index = decoded
    }

    /// Search for verses matching the query.
    /// - Supports scoped search: "BookName: keyword" limits results to that book
    /// - Multi-word queries intersect results (AND logic)
    /// - Returns VerseResult with text looked up from bibleStore
    func search(query: String, bibleStore: BibleStore, limit: Int = 50) -> [VerseResult] {
        loadIfNeeded()

        let trimmed = query.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return [] }

        var bookFilter: String? = nil
        var keywords: [String]

        // Check for scoped search with "BookName: keyword(s)"
        if let colonIndex = trimmed.firstIndex(of: ":") {
            let beforeColon = String(trimmed[trimmed.startIndex..<colonIndex]).trimmingCharacters(in: .whitespaces)
            let afterColon = String(trimmed[trimmed.index(after: colonIndex)...]).trimmingCharacters(in: .whitespaces)

            // Verify the part before the colon is a valid book name
            if !beforeColon.isEmpty && !afterColon.isEmpty && isBookName(beforeColon) {
                bookFilter = resolveBookName(beforeColon)
                keywords = afterColon.lowercased().split(separator: " ").map(String.init)
            } else {
                // Colon present but not a valid book scope — strip non-alphanumeric characters
                // so that e.g. "love: faith" becomes ["love", "faith"] instead of ["love:", "faith"]
                let cleaned = trimmed.unicodeScalars.filter { CharacterSet.alphanumerics.contains($0) || $0 == " " }
                keywords = String(cleaned).lowercased().split(separator: " ").map(String.init)
            }
        } else {
            keywords = trimmed.lowercased().split(separator: " ").map(String.init)
        }

        guard !keywords.isEmpty else { return [] }

        // Look up first keyword to get initial result set
        guard let firstMatches = index[keywords[0]] else { return [] }

        // Convert to a set of verse identifiers for intersection
        var resultSet = Set(firstMatches.map { "\($0.book).\($0.chapter).\($0.verse)" })
        var entriesByID: [String: IndexEntry] = [:]
        for entry in firstMatches {
            entriesByID["\(entry.book).\(entry.chapter).\(entry.verse)"] = entry
        }

        // Intersect with subsequent keywords (AND logic)
        for keyword in keywords.dropFirst() {
            guard let matches = index[keyword] else { return [] }
            let matchSet = Set(matches.map { "\($0.book).\($0.chapter).\($0.verse)" })
            resultSet = resultSet.intersection(matchSet)
            for entry in matches {
                let key = "\(entry.book).\(entry.chapter).\(entry.verse)"
                if entriesByID[key] == nil {
                    entriesByID[key] = entry
                }
            }
        }

        // Apply book filter if scoped search
        if let bookFilter = bookFilter {
            resultSet = resultSet.filter { id in
                guard let entry = entriesByID[id] else { return false }
                return entry.book == bookFilter
            }
        }

        // Convert to sorted array and apply limit
        let sortedIDs = resultSet.sorted { a, b in
            guard let entryA = entriesByID[a], let entryB = entriesByID[b] else { return a < b }
            let indexA = BibleStore.bookNames.firstIndex(of: entryA.book) ?? Int.max
            let indexB = BibleStore.bookNames.firstIndex(of: entryB.book) ?? Int.max
            if indexA != indexB { return indexA < indexB }
            if entryA.chapter != entryB.chapter { return entryA.chapter < entryB.chapter }
            return entryA.verse < entryB.verse
        }

        let limitedIDs = sortedIDs.prefix(limit)

        // Look up verse text from BibleStore
        return limitedIDs.compactMap { id -> VerseResult? in
            guard let entry = entriesByID[id] else { return nil }
            let verses = bibleStore.getVerses(bookName: entry.book, chapter: entry.chapter, start: entry.verse, end: entry.verse)
            let text = verses.first?.text ?? ""
            return VerseResult(book: entry.book, chapter: entry.chapter, verse: entry.verse, text: text)
        }
    }

    // MARK: - Private helpers

    /// Check if a string looks like a book name (exists in BibleStore.bookNames, case-insensitive).
    private func isBookName(_ name: String) -> Bool {
        return resolveBookName(name) != nil
    }

    /// Resolve a book name string to its canonical form.
    private func resolveBookName(_ name: String) -> String? {
        let lower = name.lowercased()
        return BibleStore.bookNames.first { $0.lowercased() == lower }
    }
}

File: src/lib/search-service.ts
import type { Verse } from "./types";
import { bookNames, loadBook, findBook as findBookName } from "./bible-store";

interface IndexEntry {
  book: string;
  chapter: number;
  verse: number;
}

export interface VerseResult {
  id: string;
  book: string;
  chapter: number;
  verse: number;
  text: string;
}

let searchIndex: Record<string, IndexEntry[]> | null = null;

async function getIndex(): Promise<Record<string, IndexEntry[]>> {
  if (searchIndex) return searchIndex;
  const data = await import("@/data/search_index.json");
  searchIndex = data.default ?? data;
  return searchIndex!;
}

export async function search(query: string, limit = 50): Promise<VerseResult[]> {
  const index = await getIndex();
  const trimmed = query.trim().toLowerCase();
  if (!trimmed) return [];

  // Scoped search: "BookName: keyword"
  let scope: string | null = null;
  let keywords: string[];
  const colonIdx = trimmed.indexOf(":");
  if (colonIdx > 0) {
    const potentialBook = trimmed.slice(0, colonIdx).trim();
    const found = findBookName(potentialBook);
    if (found) {
      scope = found;
      keywords = trimmed.slice(colonIdx + 1).trim().split(/\s+/).filter(Boolean);
    } else {
      keywords = trimmed.split(/\s+/).filter(Boolean);
    }
  } else {
    keywords = trimmed.split(/\s+/).filter(Boolean);
  }

  if (keywords.length === 0) return [];

  // AND logic: intersect results for each keyword
  let resultEntries: IndexEntry[] | null = null;
  for (const kw of keywords) {
    const entries = index[kw] ?? [];
    if (resultEntries === null) {
      resultEntries = entries;
    } else {
      const set = new Set(entries.map((e) => `${e.book}.${e.chapter}.${e.verse}`));
      resultEntries = resultEntries.filter((e) => set.has(`${e.book}.${e.chapter}.${e.verse}`));
    }
  }

  if (!resultEntries) return [];

  // Apply scope filter
  if (scope) {
    resultEntries = resultEntries.filter((e) => e.book === scope);
  }

  // Sort by Bible order
  const bookOrder = new Map(bookNames.map((n, i) => [n, i]));
  resultEntries.sort((a, b) => {
    const ba = bookOrder.get(a.book) ?? 0;
    const bb = bookOrder.get(b.book) ?? 0;
    if (ba !== bb) return ba - bb;
    if (a.chapter !== b.chapter) return a.chapter - b.chapter;
    return a.verse - b.verse;
  });

  // Limit and look up text
  const results: VerseResult[] = [];
  for (const entry of resultEntries.slice(0, limit)) {
    const book = await loadBook(entry.book);
    const chapter = book?.chapters.find((c) => c.number === entry.chapter);
    const verse = chapter?.verses.find((v) => v.number === entry.verse);
    if (verse) {
      results.push({
        id: `${entry.book}.${entry.chapter}.${entry.verse}`,
        book: entry.book,
        chapter: entry.chapter,
        verse: entry.verse,
        text: verse.text,
      });
    }
  }

  return results;
}
